### PR TITLE
add ability to get and clear sub-regions by index

### PR DESCRIPTION
**Pull request by catoverdrive**
Starting to break out changes from #6480.

This adds the ability to access sub-regions of a region by index.

@patrick-schultz I know I misnamed these---they should be child references, not parent references---I can change if you prefer, but might like to hold off until afterwards since the names of the functions are threaded through the rest of the stack :(

cc @cseed

### CODE REVIEW

**Comment by patrick-schultz** [comment]
Yeah, don't worry about naming.

File: hail/src/main/scala/is/hail/annotations/Region.scala [hunk 1]
```diff
@@ -45,13 +49,53 @@ final class Region() extends NativeBase() {
...
+
+  final def reference(other: Region): Unit = {
+    assert(explicitParents < 0, s"can't use 'reference' if you're explicitly setting Region dependencies")
+    nativeReference(other)
+  }
+
   final def refreshRegion(): Unit = nativeRefreshRegion()
+
+  def setNumParents(n: Int): Unit = {
+//    assert(nativeGetNumParents() == 0, s"Can't explicitly set Region dependencies if references already exist")
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/annotations/Region.scala]: Why is this disabled?

File: hail/src/main/c/Region.cpp [hunk 2]
```diff
@@ -88,6 +112,13 @@ void ScalaRegionPool::own(RegionPool &&pool) {
 ScalaRegion::ScalaRegion(ScalaRegionPool * pool) :
 region_(pool->pool_.get_region()) { }
 
+ScalaRegion::~ScalaRegion() {
+  region_ = nullptr;
```
[the remark this answers is not in the document]
Comment (catoverdrive) [reply hail/src/main/c/Region.cpp]: I believe it's not needed, I'll take it out.

File: hail/src/main/scala/is/hail/annotations/Region.scala (hunk 1 above)
Comment (catoverdrive) [reply hail/src/main/scala/is/hail/annotations/Region.scala]: I want to be able to expand the number of parents (but not shrink it). I'll fix the assertion, enable it, and add a test.

**Pull request merged**